File: week-2/problem-set/prob4.py
"""Prob4 example.

This file belongs to the week 2 data-structure and graph practice solutions. It uses Call, CallCenter, CustomerOrder, OrderQueue, and related helpers to illustrate the main idea with a small runnable example.
"""

import heapq

class Call:
    """Store one caller together with its priority."""
    def __init__(self, priority, caller):
        """Run the   init   step used in this example.

        Args:
            priority (int): The priority input used by this function.
            caller (str): The caller input used by this function.

        Returns:
            None: This function updates data or prints results and does not return a value.
        """
        self.priority = priority
        self.caller = caller
    
    def __lt__(self, other):
        """Run the   lt   step used in this example.

        Args:
            other (object): The other input used by this function.

        Returns:
            bool: True when the required condition is satisfied, otherwise False.
        """
        return self.priority < other.priority

class CallCenter:
    """Priority queue model for a call centre."""
    def __init__(self):
        """Run the   init   step used in this example.

        Returns:
            None: This function updates data or prints results and does not return a value.
        """
        self.calls = []
    
    def add_call(self, priority, caller):
        """Insert a new call into the priority queue.

        Args:
            priority (int): The priority input used by this function.
            caller (str): The caller input used by this function.

        Returns:
            None: This function updates data or prints results and does not return a value.
        """
        heapq.heappush(self.calls, Call(priority, caller))
    
    def answer_call(self):
        """Remove and return the next caller to answer.

        Returns:
            object: The value produced by this function.
        """
        if self.calls:
            return heapq.heappop(self.calls).caller
        else:
            return None

class CustomerOrder:
    """Store one customer order."""
    def __init__(self, order_number, customer):
        """Run the   init   step used in this example.

        Args:
            order_number (object): The order_number input used by this function.
            customer (str): The customer input used by this function.

        Returns:
            None: This function updates data or prints results and does not return a value.
        """
        self.order_number = order_number
        self.customer = customer

class OrderQueue:
    """First-in, first-out queue for customer orders."""
    def __init__(self):
        """Run the   init   step used in this example.

        Returns:
            None: This function updates data or prints results and does not return a value.
        """
        self.orders = []
    
    def add_order(self, order_number, customer):
        """Add a customer order to the back of the queue.

        Args:
            order_number (object): The order_number input used by this function.
            customer (str): The customer input used by this function.

        Returns:
            None: This function updates data or prints results and does not return a value.
        """
        self.orders.append(CustomerOrder(order_number, customer))
    
    def send_order(self):
        """Send the next order in first-in, first-out order.

        Returns:
            object: The value produced by this function.
        """
        if self.orders:
            return self.orders.pop(0).customer
        else:
            return None

class Calculator:
    """Postfix-expression calculator built on a stack."""
    def __init__(self):
        """Run the   init   step used in this example.

        Returns:
            None: This function updates data or prints results and does not return a value.
        """
        self.stack = []
    
    def calculate(self, expression):
        """Evaluate a postfix expression using a stack.

        Args:
            expression (str): The expression input used by this function.

        Returns:
            int: The integer result produced by the algorithm.
        """
        stack = []

        for token in expression.split():
            if token.isdigit():
                stack.append(int(token))
            else:
                operand2 = stack.pop()
                operand1 = stack.pop()
                if token == '+':
                    stack.append(operand1 + operand2)
                elif token == '-':
                    stack.append(operand1 - operand2)
                elif token == '*':
                    stack.append(operand1 * operand2)
                elif token == '/':
                    stack.append(int(operand1 / operand2))
                else:
                    raise ValueError('Invalid operator')

        if len(stack) != 1:
            raise ValueError('Invalid expression')

        return stack.pop()



# Create a CallCenter object
cc = CallCenter()

# Add some calls to the queue with different priorities
cc.add_call(3, "John")
cc.add_call(1, "Jane")
cc.add_call(2, "Jim")

# Answer the highest priority call
print(cc.answer_call())  # "Jane"
print(cc.answer_call())  # "Jim"
print(cc.answer_call())  # "John"
print(cc.answer_call())  # None (no more calls in queue)

# Create an OrderQueue object
oq = OrderQueue()

# Add some orders to the queue
oq.add_order(1, "John")
oq.add_order(2, "Jane")
oq.add_order(3, "Jim")

# Send the next order in the queue
print(oq.send_order())  # "John"
print(oq.send_order())  # "Jane"
print(oq.send_order())  # "Jim"
print(oq.send_order())  # None (no more orders in queue)

# Create a Calculator object
calc = Calculator()

result = calc.calculate('5 1 2 + 4 * + 3 -') # Evaluates to 14
"""
In this example, we evaluate the expression "5 1 2 + 4 * + 3 -", 
which represents the infix expression "(5 + (1 + 2) * 4) - 3". 
The result should be 14, and this is what is printed to the console.
"""
print(result)

File: week-2/problem-set/prob4.cpp
/**
 * @file prob4.cpp
 * @brief C++ version of the call-centre, order-queue, and postfix-calculator example.
 */

#include <iostream>
#include <queue>
#include <stdexcept>
#include <string>
#include <vector>

using namespace std;

struct Call {
    int priority;
    string caller;

    /**
     * @brief Reverse the comparison so smaller priority numbers are served first.
     * @param other const Call& Another call record used by the priority queue.
     * @return bool True when this call should come after other inside the heap structure.
     */
    bool operator<(const Call& other) const {
        return priority > other.priority;
    }
};

class CallCenter {
public:
    /**
     * @brief Add a new call to the priority queue.
     * @param priority int Priority number of the caller. Smaller values mean more urgent calls.
     * @param caller const string& Name of the caller.
     * @return void This method updates the queue and does not return a value.
     */
    void add_call(int priority, const string& caller) {
        calls.push(Call{priority, caller});
    }

    /**
     * @brief Answer the next call.
     * @return string Name of the caller who should be served next, or an empty string when no calls remain.
     */
    string answer_call() {
        if (calls.empty()) {
            return "";
        }
        string caller = calls.top().caller;
        calls.pop();
        return caller;
    }

private:
    priority_queue<Call> calls;
};

struct CustomerOrder {
    int order_number;
    string customer;
};

class OrderQueue {
public:
    /**
     * @brief Add a new customer order to the back of the queue.
     * @param order_number int Identifier of the order.
     * @param customer const string& Name of the customer who placed the order.
     * @return void This method updates the queue and does not return a value.
     */
    void add_order(int order_number, const string& customer) {
        orders.push(CustomerOrder{order_number, customer});
    }

    /**
     * @brief Send the next order using first-in, first-out order.
     * @return string Name of the customer whose order is sent next, or an empty string when no orders remain.
     */
    string send_order() {
        if (orders.empty()) {
            return "";
        }
        string customer = orders.front().customer;
        orders.pop();
        return customer;
    }

private:
    queue<CustomerOrder> orders;
};

class Calculator {
public:
    /**
     * @brief Evaluate a postfix expression with integer arithmetic.
     * @param expression const string& Space-separated postfix expression.
     * @return int Computed integer result of the expression.
     */
    int calculate(const string& expression) {
        vector<int> stack;
        string token;

        for (char character : expression) {
            if (character == ' ') {
                if (!token.empty()) {
                    process_token(token, stack);
                    token.clear();
                }
            } else {
                token.push_back(character);
            }
        }
        if (!token.empty()) {
            process_token(token, stack);
        }
        if (stack.size() != 1) {
            throw runtime_error("Invalid expression");
        }
        return stack.back();
    }

private:
    /**
     * @brief Process one token from the postfix expression.
     * @param token const string& Current token, either a number or an operator.
     * @param stack vector<int>& Evaluation stack used by the postfix algorithm.
     * @return void This helper updates the stack and does not return a value.
     */
    void process_token(const string& token, vector<int>& stack) {
        if (isdigit(static_cast<unsigned char>(token[0]))) {
            stack.push_back(stoi(token));
            return;
        }

        int operand2 = stack.back();
        stack.pop_back();
        int operand1 = stack.back();
        stack.pop_back();

        if (token == "+") {
            stack.push_back(operand1 + operand2);
        } else if (token == "-") {
            stack.push_back(operand1 - operand2);
        } else if (token == "*") {
            stack.push_back(operand1 * operand2);
        } else if (token == "/") {
            stack.push_back(operand1 / operand2);
        } else {
            throw runtime_error("Invalid operator");
        }
    }
};

/**
 * @brief Run the queue and calculator demonstrations.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    CallCenter cc;
    cc.add_call(3, "John");
    cc.add_call(1, "Jane");
    cc.add_call(2, "Jim");
    cout << cc.answer_call() << '\n';
    cout << cc.answer_call() << '\n';
    cout << cc.answer_call() << '\n';
    cout << cc.answer_call() << '\n';

    OrderQueue oq;
    oq.add_order(1, "John");
    oq.add_order(2, "Jane");
    oq.add_order(3, "Jim");
    cout << oq.send_order() << '\n';
    cout << oq.send_order() << '\n';
    cout << oq.send_order() << '\n';
    cout << oq.send_order() << '\n';

    Calculator calc;
    cout << calc.calculate("5 1 2 + 4 * + 3 -") << '\n';
    return 0;
}
